flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistry.java

```diff
@@ -0,0 +1,192 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.state;
+
+import org.apache.flink.annotation.VisibleForTesting;
+import org.apache.flink.api.java.tuple.Tuple2;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A {@code SharedStateRegistry} will be deployed in the 
+ * {@link org.apache.flink.runtime.checkpoint.CheckpointCoordinator} to 
+ * maintain the reference count of those state objects shared among different
+ * checkpoints. Each shared state object must be identified by a unique key. 
+ */
+public class SharedStateRegistry implements Serializable {
```

Comment:
I made it serializable due to the serialization issues in powermock. In some tests for `CompletedCheckpointStore`, I mock some `SubtaskState`s and attempt to verify that they can correctly register their shared states when the `CompletedCheckpointStore` is recovered. To allow the storage of the tests work, all classes related to the mocked objects should be serializable. 

With your changes made in the storage of `CompletedCheckpoint`, I think we can make the tests work without requiring `SharedStateRegistry` to be serializable. But at prior to that, i prefer to keep `SharedStateRegistry` serializable so that the tests can correctly work now.